File: other/past202004-open/N.py
import sys
from operator import itemgetter, add

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


class DualSegmentTree:
    def __init__(self, size: int, segfunc, lazy_ide_ele=0):
        self.lazy_ide_ele = lazy_ide_ele
        self.segfunc = segfunc
        self.N0 = 1 << (size - 1).bit_length()
        self.lazy = [self.lazy_ide_ele] * (2 * self.N0)
    
    def gindex(self, left, right):
        L = left + self.N0
        R = right + self.N0
        lm = (L // (L & -L)) >> 1
        rm = (R // (R & -R)) >> 1
        while L < R:
            if R <= rm:
                yield R
            if L <= lm:
                yield L
            L >>= 1
            R >>= 1
        while L:
            yield L
            L >>= 1
    
    def propagates(self, *ids):
        for i in reversed(ids):
            idx = i - 1
            v = self.lazy[idx]
            if v == self.lazy_ide_ele:
                continue
            self.lazy[2 * idx + 1] = self.segfunc(self.lazy[2 * idx + 1], v)
            self.lazy[2 * idx + 2] = self.segfunc(self.lazy[2 * idx + 2], v)
            self.lazy[idx] = self.lazy_ide_ele
    
    def update(self, left: int, right: int, x):
        L = self.N0 + left
        R = self.N0 + right
        
        while L < R:
            if R & 1:
                R -= 1
                self.lazy[R - 1] = self.segfunc(self.lazy[R - 1], x)
            if L & 1:
                self.lazy[L - 1] = self.segfunc(self.lazy[L - 1], x)
                L += 1
            L >>= 1
            R >>= 1
    
    def query(self, k: int):
        self.propagates(*self.gindex(k, k + 1))
        return self.lazy[k + self.N0 - 1]


def solve():
    N, Q = map(int, rl().split())
    events = []
    y = set()
    for _ in range(N):
        xmin, ymin, d, c = map(int, rl().split())
        events.append((0, xmin, ymin, ymin + d, c))
        events.append((2, xmin + d, ymin, ymin + d, c))
        y.add(ymin)
        y.add(ymin + d)
    for i in range(Q):
        a, b = map(int, rl().split())
        events.append((1, a, b, i))
        y.add(b)
    
    events.sort(key=itemgetter(1, 0))
    y_to_idx = {val: idx for idx, val in enumerate(sorted(y))}
    
    dual_seg_tree = DualSegmentTree(len(y), add)
    ans = [0] * Q
    for com, *event in events:
        if com == 0:
            _, ys, ye, cost = event
            dual_seg_tree.update(y_to_idx[ys], y_to_idx[ye] + 1, cost)
        elif com == 2:
            _, ys, ye, cost = event
            dual_seg_tree.update(y_to_idx[ys], y_to_idx[ye] + 1, -cost)
        else:
            _, bi, qi = event
            ans[qi] = dual_seg_tree.query(y_to_idx[bi])
    print(*ans, sep='\n')


if __name__ == '__main__':
    solve()

File: other/past202004-open/N.nim
import algorithm, deques, math, sets, sequtils, strutils, sugar, tables


proc bit_length(n: int): int =
    if n == 0:
      return 0
    let s = toBin(n, 60)
    return 60 - s.find('1')


type
    DualSegmentTree*[T] = ref object
        LV, N0: Natural
        lazy_ide_ele: T
        lazy_data: seq[T]

proc initDualSegmentTree*[T](size: Natural, lazy_ide_ele: T): DualSegmentTree[T] =
    var
        LV = bit_length(size - 1)
        N0 = 1 shl LV
    return DualSegmentTree[T](LV: LV, N0: N0, lazy_ide_ele: lazy_ide_ele, lazy_data: newSeqWith(2 * N0, lazy_ide_ele))

iterator gindex*[T](self: var DualSegmentTree[T], left, right: int): int =
    var
        L = (left + self.N0) shr 1
        R = (right + self.N0) shr 1
        lc = if (left and 1) == 1: 0 else: bit_length(L and -L)
        rc = if (right and 1) == 1: 0 else: bit_length(R and -R)
    for i in 0..<self.LV:
        if rc <= i:
            yield R
        if L < R and lc <= i:
            yield L
        L = L shr 1
        R = R shr 1

proc propagates*[T](self: var DualSegmentTree[T], ids: seq[int]) =
    var
        idx: int
        v: T
    for id in reversed(ids):
        idx = id - 1
        v = self.lazy_data[idx]
        if v == self.lazy_ide_ele:
            continue
        self.lazy_data[2*idx + 1] += v
        self.lazy_data[2*idx + 2] += v
        self.lazy_data[idx] = self.lazy_ide_ele

proc update*[T](self: var DualSegmentTree[T], left, right: int, x: T) =
    var
        L = left + self.N0
        R = right + self.N0
    
    while L < R:
        if (L and 1) == 1:
            self.lazy_data[L - 1] += x
            inc L
        if (R and 1) == 1:
            dec R
            self.lazy_data[R - 1] += x
        L = L shr 1
        R = R shr 1

proc query*[T](self: var DualSegmentTree[T], k: int): T =
    self.propagates(toSeq(self.gindex(k, k + 1)))
    return self.lazy_data[k + self.N0 - 1]


type
    Event = array[5, int]

proc eventCmp(A, B: Event): int =
    if A[1] < B[1]: return -1
    if A[1] > B[1]: return 1
    if A[0] < B[0]: return -1
    if A[0] > B[0]: return 1
    else: return 0


var
    events: seq[Event]
    y: HashSet[int]
    y_to_idx: Table[int, int]
    dual_seg_tree: DualSegmentTree[int]
    ans: seq[int]


proc solve() =
    var N, Q: int
    (N, Q) = stdin.readLine.split.map(parseInt)
    events = newSeq[Event]()
    y = initHashSet[int]()
    var xmin, ymin, d, c, a, b: int
    for _ in 0..<N:
        (xmin, ymin, d, c) = stdin.readLine.split.map(parseInt)
        events.add([0, xmin, ymin, ymin + d, c])
        events.add([2, xmin + d, ymin, ymin + d, c])
        y.incl(ymin)
        y.incl(ymin + d)
    for i in 0..<Q:
        (a, b) = stdin.readLine.split.map(parseInt)
        events.add([1, a, b, i, 0])
        y.incl(b)
    
    events.sort(eventCmp)
    y_to_idx = initTable[int, int]()
    for idx, val in sorted(toSeq(y.items)):
        y_to_idx[val] = idx
    
    dual_seg_tree = initDualSegmentTree(y.len, 0)
    ans = newSeq[int](Q)
    var ys, ye, cost, bi, i: int
    for event in events:
        if event[0] == 0:
            (ys, ye, cost) = event[2..^1]
            dual_seg_tree.update(y_to_idx[ys], y_to_idx[ye] + 1, cost)
        elif event[0] == 2:
            (ys, ye, cost) = event[2..^1]
            dual_seg_tree.update(y_to_idx[ys], y_to_idx[ye] + 1, -cost)
        else:
            (bi, i) = event[2..^2]
            ans[i] = dual_seg_tree.query(y_to_idx[bi])
    echo ans.join("\n")


when is_main_module:
    solve()
